spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -637,20 +662,63 @@ meaning of the term `eventual consistency` as used in this specification).
 participant has been successfully notified.
 
-The LRA model is quite flexible in how business methods perform compensations.
-If a compensating activity is brief then the synchronous model may be appropriate:
+The LRA model supports both synchronous and asynchronous programming models.
+For example, if a compensating activity is brief then the synchronous model may
+be appropriate as the following code snippet shows. The snippet also shows how
+the compensating code can check for a <<nesting-lras,nested context>>):
 
 [source,java]
 ----
     @Compensate
     @Path("/compensate")
     @PUT
-    public Response compensatePreviousAction(
-            @HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
-        getActivity(lraId).setCompensated();  // business logic for handling the LRA, not provided by spec or implementation.
+    public Response compensatePreviousAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId,
+                                             @HeaderParam(LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentLRA) {
+        if (parentLRA != null) 
+            // A nested context is being cancelled. Nested contexts are explained in detail in a later section.
+
+            // Code which is sensitive to executing with a nested context goes here. Generally
+            // only the @Complete handler is sensitive to executing with a nested context.
+        }
+
+        // perform business logic which compensates for any work performed in the context of the LRA
+        getActivity(lraId).setCompensated();
+
+        return Response.ok();
+    }
+
+    @Complete
+    @Path("/complete")
+    @PUT
+    public Response cleanupPreviousAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId,
+                                          @HeaderParam(LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentLRA) {
+        if (parentLRA != null) 
+            // A nested context is being cancelled. Nested contexts are explained in detail in a later section.
+
+            // If a nested context is closed but the parent context is later cancelled
+            // then any work performed in the nested context must still be capable of
+            // being compensated for. Therefore, even though no more work will be executed
+            // with the nested context, the business logic must remember how to compensate
+            // for any work it did do (until explicitly told to forget about the nested context).
+        }
+
+        // business logic for handling the LRA (provided by the business developer)
+        getActivity(lraId).setCompensated();
 
         return Response.ok();
     }
+
+    @Forget
+    @Path("/forget")
+    @DELETE
+    public Response forget(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId,) {
+                           @HeaderParam(LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentLRA) {
+        if (parentLRA != null) {
+            // code which is sensitive to executing with a nested context goes here
+            ...
```

Comment:
also please mention that you should clean up the information required for possible nested compensation here.

Reply:
I cleaned up all of this example so if you don't mind can you recheck the commentary I added to the code snippets.